숫자게임 [STEP 2] kaki, Goat, 혜모리

kak1x opened the pull request:
@yuvinrho 
안녕하세요 로빈! 숫자야구 마지막 스탭입니다. 리뷰 잘 부탁드리겠습니다!

### 고민했던 점
1. 어떤 방식으로 유효성 검증을 해야할지 고민했습니다.
    - 메뉴 입력 : 1 또는 2가 아닐 경우 에러 처리
    - 사용자 번호 입력 : 양수(`UInt` 타입)와 배열의 숫자 크기 비교(9 초과 시), 입력 받은 배열 내 인덱스 개수가 3이 아니면 에러 문구 출력
2. 에러를 다루는 방법을 고민했습니다.
    - `throw` + `do` + `catch` 구문 사용
3. guard let 처리 시 제어 전달문(`Control Transfer Statement`)의 사용

### 새로 알게된 점
1. UInt 타입 사용 (양수만 골라내기)
2. 에러 처리 방법 (`if`, `do-catch`)

### 조언이 필요한 점
1. `startGame` 함수의 `do-catch`문 에러 처리를 좀 더 깔끔하게 할 수 있을까요?
2. 사용자 입력 시 에러처리는 에러 후에도 계속 게임을 진행해야해서 `do-catch`문을 사용하지 못했습니다. 이 부분에 대해 조언을 얻고 싶습니다.

hyemory committed:
feat: menuInputOutput 함수 추가

Goatt8 committed:
feat: inputNumbers 함수구현

kak1x committed:
feat: 메뉴 선택 에러처리 추가

Goatt8 committed:
feat: 사용자 입력 유효성 검증

hyemory committed:
feat: 게임 종료 구현

kak1x committed:
feat: STEP 1 수정사항 반영

hyemory committed:
docs: 순서도 수정

hyemory committed:
Merge branch 'ic_8_kaki1' into step2

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,22 +1,27 @@
-// 
+//
 
 import Foundation
```
`import Foundation`은 뭘까요? 이게 없으면 될까요 안될까요? 
추가적으로 공부해보시면 좋을것 같아서 알려드립니다☺️

kak1x replied:
Foundation 프레임워크를 추가해주는 구문인데, 다른건 잘 몰라도 68번째 줄에 있는 components 메서드를 이용하기 위해선 Foundation이 필요합니다. components는 Foundation 프레임워크에 속해 있기 때문입니다!

yuvinrho replied:
👍 맞습니다~ components 를 사용하려면 필요합니다. components 이외 말고 다른것들은 뭐가 있는지 봐두시면 나중에 프로젝트 진행하실 때 좋을것 같습니다~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,22 +1,27 @@
+    case invalidNumber
+}
+
+var winNumberArray: [UInt] = []
+var userNumberArray: [UInt] = []
 var strikeNumber = 0
 var ballNumber = 0
 var chance = 9
+var menuNumber = 0
```
전역변수로 사용한 이유가 있을까요?
[공식문서 Global and Local Variables](https://docs.swift.org/swift-book/LanguageGuide/Properties.html#//apple_ref/doc/uid/TP40014097-CH14-XID_382) 를 참고해서
전역변수와 지역변수의 차이를 공부해보고 어떤걸 사용하면 좋을지 고민해보면 어떨까요?

kak1x replied:
지역변수로 선언해주면 여러 함수에서 변수를 사용하기 어려워서 전역변수로 선언하여 사용하였습니다. 초기화를 계속 해주는것도 좀 지저분하다고 생각했었는데, 지역변수로 활용할 방법을 생각해보겠습니다! 아마 함수에 파라미터를 설정해주면 바꿀 수 있을 듯 싶네요

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,22 +1,27 @@
 var strikeNumber = 0
 var ballNumber = 0
 var chance = 9
+var menuNumber = 0
 
-func createRandomNumbers() -> [Int] {
-    var randomNumbers: Set<Int> = []
+func createRandomNumbers() -> [UInt] {
+    var randomNumbers: Set<UInt> = []
     
     while randomNumbers.count < 3 {
-        randomNumbers.insert(Int.random(in: 1...9))
+        randomNumbers.insert(UInt.random(in: 1...9))
```
타입을 Int에서 UInt로 수정하신 이유가 뭔지 궁금합니다~ (틀렸다는 것이 아니고 궁금해서 그렇습니다☺️)

kak1x replied:
음수의 유효성 검증을 따로 해주지 않아도 되어 UInt로 바꿨던 기억이 있습니다! 근데 0도 포함이 될 줄 몰랐어서 이 부분은 수정이 필요할 것 같네요.

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -32,19 +37,85 @@ func countStrikeBall() {
 }
 
-func playGame() {
+func menuInputOutput() throws {
```
`menuInputOutput()` 함수명이 어색한거 같아요~  [API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/) 를 참고해서 한번 더 고민해보면 어떨까요?

kak1x replied:
고민해보도록 하겠습니다.

hyemory replied:
`choiceMenu`로 이름 변경했습니다.

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -32,19 +37,85 @@ func countStrikeBall() {
+    
+    let inputMenu = readLine()
+    
+    guard let menu = inputMenu else {
+        throw InputError.invalidNumber
+    }
+    
+    guard menu == "1" || menu == "2" else {
+        throw InputError.invalidNumber
+    }
```
guard 문을 사용해서 옵셔널을 처리하신 부분 너무 좋습니다👍
위, 아래 코드가 같은데 해결할 방법이 있을까요?

kak1x replied:
and로 이어주면 합칠 수 있을 것 같습니다!

hyemory replied:
메뉴 타입을 새로 정의하여 해결했습니다

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -32,19 +37,85 @@ func countStrikeBall() {
+    let menuIntOptional = Int(menu)
+    
+    if let menuInt = menuIntOptional {
+        menuNumber = menuInt
```
메뉴 넘버를 전역변수로 넘겨주는 것은 어색?하게 보입니다~
위에 남겨드린 코멘트와 같이 어떤것을 전역으로 쓸지 지역으로 쓸지 고민해보면 좋을 것 같습니다~

kak1x replied:
넵 알겠습니다

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -32,19 +37,85 @@ func countStrikeBall() {
+        return inputNumbers()
+    }
+    
+    let inputNumberArray = inputNumber.components(separatedBy: " ").map{ UInt($0) }
```
`inputNumberArray` 에 Array가 안들어가도 될것 같아요~ 이것도 [API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/) 를 참고해보시면 좋을 것 같습니다~

코드 13, 14 줄 변수명 Array 붙어있는 것도 같이 확인해주세요~

yuvinrho replied:
고차함수 `map` 을 사용해보신 점 너무 좋습니다👍
[compactMap](https://developer.apple.com/documentation/swift/sequence/compactmap(_:))도 추가적으로 공부해보면 좋을 것 같습니다~

kak1x replied:
참고해보겠습니다. 이후에 알게됐는데 compactMap이 되게 간단하고 좋은 것 같더라구요. 이것도 공부해서 사용해볼게요!

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -32,19 +37,85 @@ func countStrikeBall() {
+        return inputNumbers()
+    }
+    
+    let inputNumberArray = inputNumber.components(separatedBy: " ").map{ UInt($0) }
+    
+    for number in inputNumberArray {
+        if let number = number {
+            userNumberArray.append(number)
+        }
+    }
+    
+    guard userNumberArray.count == 3 else {
+        print("입력이 잘못되었습니다")
+        userNumberArray = []
+        return inputNumbers()
+    }
+    
+    if userNumberArray[0] > 9 {
+        print("입력이 잘못되었습니다")
+        userNumberArray = []
+        return inputNumbers()
+    } else if userNumberArray[1] > 9 {
+        print("입력이 잘못되었습니다")
+        userNumberArray = []
+        return inputNumbers()
+    } else if userNumberArray[2] > 9 {
+        print("입력이 잘못되었습니다")
+        userNumberArray = []
+        return inputNumbers()
+    }
+    
+    let userNumberSet = Set(userNumberArray)
+    let userNumbers = Array(userNumberSet)
+    
+    guard userNumbers.count == 3 else {
+        print("입력이 잘못되었습니다")
+        userNumberArray = []
+        return inputNumbers()
+    }
+}
```
이 함수는 사용자 입력에 대한 에러처리를 해보면 어떨까요?

yuvinrho replied:
숫자가 1-9 까지 들어올 수 있는데 0도 가능합니다 수정해주세요~
![image](https://user-images.githubusercontent.com/49301866/209148377-b69dd345-37a9-4014-90c8-693290416477.png)

kak1x replied:
확인해보겠습니다.

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -32,19 +37,85 @@ func countStrikeBall() {
+        return inputNumbers()
+    }
+    
+    if userNumberArray[0] > 9 {
+        print("입력이 잘못되었습니다")
+        userNumberArray = []
+        return inputNumbers()
+    } else if userNumberArray[1] > 9 {
+        print("입력이 잘못되었습니다")
+        userNumberArray = []
+        return inputNumbers()
+    } else if userNumberArray[2] > 9 {
+        print("입력이 잘못되었습니다")
+        userNumberArray = []
+        return inputNumbers()
+    }
```
이 부분 다른방법으로 처리할 수 있을까요??

kak1x replied:
or로 묶어서 처리 가능할 것 같습니다. 혹은 for문을 돌려볼 수도 있겠네요.

hyemory replied:
~= 연산자를 사용하여 한 줄로 처리했습니다.

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -53,11 +124,32 @@ func playGame() {
+    }
+}
+
+func startGame() throws{
```
`throws` 키워드는 언제 사용하는 걸까요?
이 함수는 `throws` 가 필요할까요?
[Error Handling](https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html)를 참고해보면 좋을것 같습니다~

yuvinrho replied:
throws와 { 가 붙어있어요~

kak1x replied:
에러 핸들링이 잘 기억이 안나서.. 다시 공부해볼게요 ㅠㅠ

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -32,19 +37,85 @@ func countStrikeBall() {
+        throw InputError.invalidNumber
+    }
+    
+    guard menu == "1" || menu == "2" else {
```
메뉴를 "1" , "2" 말고 조금 더 직관성 있게 만들 수 있지 않을까요?
시간되시면 [Enum](https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html)과 magic number, masic literal에 대해 공부해보시는 것도 좋을 것 같습니다~

kak1x replied:
넵 공부해보도록 하겠습니다!

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -53,11 +124,32 @@ func playGame() {
 }
 
-playGame()
+try startGame()
```
`try` 키워드는 언제 사용하는 걸까요?
위에 남긴 코멘트 `throws` 랑 같이 [Error Handling](https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html)를 공부해보시면 좋을 것 같습니다~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -53,11 +124,32 @@ func playGame() {
+        chance = 9
         strikeNumber = 0
         ballNumber = 0
+        userNumberArray.removeAll()
```
변수를 여기에 초기화해주는게 어색해 보입니다~ 혹시 다른 방법은 없을까요?
이것도 위에 코멘트에  지역변수, 전역변수 관련해서 해결가능하실 것 같습니다~

hyemory replied:
지역 변수로 수정했습니다.

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -32,19 +37,85 @@ func countStrikeBall() {
+    }
+}
+
+func playGame() throws {
```
`throws` 키워드를 사용한 이유가 있으신가요?

kak1x replied:
이 부분도 에러 핸들링을 다시 봐보겠습니다.

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -53,11 +124,32 @@ func playGame() {
+    }
+    catch InputError.invalidNumber {
+        print("입력이 잘못되었습니다")
+        try startGame()
```
여기도 `try` ~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -53,11 +124,32 @@ func playGame() {
+        if menuNumber == 2 {
+            return
+        }
+        try playGame()
```
`try` 여기도 있네요~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -32,19 +37,85 @@ func countStrikeBall() {
         
         if strikeNumber == 3 {
             print("사용자 승리!")
+            try startGame()
```
`try` 키워드가 있네요?

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -53,11 +124,32 @@ func playGame() {
         }
         else {
             print("컴퓨터 승리...!")
+            try startGame()
```
여기도 `try`가 있군요~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,22 +1,27 @@
+
+var winNumberArray: [UInt] = []
+var userNumberArray: [UInt] = []
 var strikeNumber = 0
 var ballNumber = 0
```
해당 변수는 개수를 나타내는 값이기 때문에 `Number` 보다 `Count`가 더 낫지 않을까라는 생각이 드는데 어떻게 생각하시나요?

kak1x replied:
Count가 조금 더 어울리는 이름 같네요

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,22 +1,27 @@
-        randomNumbers.insert(Int.random(in: 1...9))
+        randomNumbers.insert(UInt.random(in: 1...9))
     }
     
     return Array(randomNumbers)
```
여기는 리턴 위에 한 줄 띄어져있고 나머지는 다 붙어있네요~
개행을 하는 것과 안하는 것 중 어떤게 더 가독성이 좋을지 고민해보고 컨벤션을 통일하면 좋을 것 같습니다~

kak1x replied:
앗 이 부분도 수정하도록 하겠습니다

yuvinrho reviewed (changes requested):
안녕하세요 여러분~~ 마지막 스텝까지 고생 많으셨습니다!🎉 마지막 리뷰라니 아쉽네요😭
이번에 브랜치 충돌 났는데 너무 잘 해결해주셨습니다!! 이후에도 충돌이 발생하면 잘 해결하시리라 믿습니다!
step1에서 말씀드린 코드 스타일 부분 많이 깔끔해진 것 같아서 보기 훨씬 좋았습니다. 기능분리도 잘해주셨어요! 최고최고👍👍
조금 더 개선하면 좋을 것 같은 부분 코멘트를 남겨두었으니 확인부탁드립니다~
코멘트 이해 안가신 부분은 언제든지 디엠남겨 주세요~

kak1x committed:
feat: STEP 2 수정 사항 반영

1. 전역변수 제거
2. 변수명, 함수명 변경
3. UInt 타입 Int 타입으로 변경
4. 사용자 입력 유효성 검사 방법 변경
5. 메뉴 종류 정의 및 메뉴 기능 선택 방법 변경

kak1x committed:
feat: 불필요한 변수 제거 및 startGame 함수 수정

hyemory committed:
docs: README 프로젝트 회고 작성

hyemory committed:
feat: 사용자 입력 오류 처리 수정

Goatt8 committed:
docs: 팀회고 작성

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
+    case exit = 2
+}
+
+var winNumberList: [Int] = []
+var userNumberList: [Int] = []
```
굳굳👍👍 이것도 좋고 개인적으로 winNumbers, userNumbers도 괜찮았을 것 같습니다~ (수정안하셔도 돼요 개인적인 의견입니다!)

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
+    case invalidNumberOfUser
+}
+
+enum Menu: Int {
+    case start = 1
+    case exit = 2
+}
```
굳굳 열거형을 사용해보신 점 좋습니다~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
     while randomNumbers.count < 3 {
         randomNumbers.insert(Int.random(in: 1...9))
     }
-    
```
저는 개인적으로 리턴문을 쓸때 개행을 하는 편 입니다☺️ 애플에서 제공해주는 공식 문서나 예제는 어떻게 쓰였는지, 다른 사람은 어떻게 했는지 등 여러 코드를 보고 고민하면서 여러분들의 컨벤션을 만들어 나가면 좋을 것 같습니다👍

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
-            print("남은기회 : \(chance)")
+    let inputMenu = readLine()
+    
+    guard let menuString = inputMenu else {
+        throw InputError.invalidNumberOfMenu
+    }
+    
+    guard let menuRawValue = Int(menuString) else {
+        throw InputError.invalidNumberOfMenu
+    }
```
이 부분은 합쳐보실 수 있을것 같습니다~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
-            strikeNumber += 1
-        } else if winNumberArray.contains(userNumberArray[index]) {
-            ballNumber += 1
+func countStrikeBall(strikeCount: inout Int, ballCount: inout Int) {
```
인아웃을 사용해서 해결해보신 점 좋습니다! 👍👍
다른 방법이 있다면 튜플로 스트라이크, 볼 카운트를 반환하는 방법도 있을것 같습니다~ (수정X, 현재 방법도 좋은 방법인것 같습니다!)
이건 나중에 [공식 스위프트 문서](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html)에 튜플 부분도 한번 공부해보세요~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
+        throw InputError.invalidNumberOfUser
+    }
+    
+    userNumberList = inputNumber.components(separatedBy: " ").compactMap{ Int($0) }
```
오! 이번에는 compactMap을 사용해보셨네요!!👍👍 고차함수는 참 좋은 기능인 것 같습니다~
[compactMap](https://developer.apple.com/documentation/swift/sequence/compactmap(_:))에 리턴타입은 무엇이고 그러면 `userNumberList`는 리턴타입은 어떻게되나요?

kak1x replied:
compactMap 리턴타입은 nil을 제외한 배열이고, userNumberList는 Int 배열을 리턴하게 될 것 같습니다!

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
+    
+    userNumberList = inputNumber.components(separatedBy: " ").compactMap{ Int($0) }
+    
+    guard userNumberList.count == 3 else {
+        userNumberList = []
+        throw InputError.invalidNumberOfUser
+    }
+    
+    guard (1...9) ~= userNumberList[0] && (1...9) ~= userNumberList[1] && (1...9) ~= userNumberList[2]  else {
+        userNumberList = []
+        throw InputError.invalidNumberOfUser
+    }
+    
+    let userNumberSet = Set(userNumberList)
+    let userNumbers = Array(userNumberSet)
+    
+    guard userNumbers.count == 3 else {
+        userNumberList = []
+        throw InputError.invalidNumberOfUser
+    }
```
이 부분은 중복된 숫자를 체크하기 위함이 맞으실까요?
그렇다면 65~68줄이랑 같이 해볼 수 있을것 같습니다☺️

yuvinrho replied:
guard 는 어떨때 사용하나요? 위에 코멘트 compactMap 리턴타입과 같이 보시면 될것 같습니다~
그리고 guard문 else 블록에 `userNumberList = []` 가 있어야될까요? 한번 더 고민해봅시다☺️

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
+        throw InputError.invalidNumberOfUser
+    }
+    
+    guard (1...9) ~= userNumberList[0] && (1...9) ~= userNumberList[1] && (1...9) ~= userNumberList[2]  else {
```
~= 연산자를 사용하면 범위에 맞는지 확인할 수 있군요!!  저두 나중에 저 연산자를 사용해봐야겠어요 굳굳👍

yuvinrho replied:
인덱스를 각 각 주셨는데 더 좋은 방법으로 처리하실 수 있을것 같아요~ 스텝1에서 해보셨으니 금방 하실 것 같습니다!

kak1x replied:
반복문 말씀하시는거 맞나요?!

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
+        switch Menu(rawValue: try choiceMenu()) {
+        case .start:
+            var chance = 9
+            var strikeCount = 0
+            var ballCount = 0
```
이 변수들이 여기에 필요할까요? 그럼 이 변수들은 어디에 위치하면 좋을까요?

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
+            var chance = 9
+            var strikeCount = 0
+            var ballCount = 0
+            userNumberList.removeAll()
```
127번줄은 userNumberList가 전역변수로 사용되지 않으면 사라질 것 같네요~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
+            var strikeCount = 0
+            var ballCount = 0
+            userNumberList.removeAll()
+            playGame(chance: &chance, strikeCount: &strikeCount, ballCount: &ballCount)
```
위에 strikeCount와 ballCount 변수 위치에 따라 매개변수가 달라질 수 있겠습니다~

yuvinrho commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,63 +1,142 @@
+func startGame() {
+    do {
+        switch Menu(rawValue: try choiceMenu()) {
+        case .start:
+            var chance = 9
+            var strikeCount = 0
+            var ballCount = 0
+            userNumberList.removeAll()
+            playGame(chance: &chance, strikeCount: &strikeCount, ballCount: &ballCount)
+        case .exit:
+            return
+        default:
+            throw InputError.invalidNumberOfMenu
         }
```
switch 를 사용해서 메뉴 분기처리를 해보신 점 너무 좋습니다👍
